src/generator.rs

```diff
@@ -217,8 +231,109 @@ impl<'a> Generator<'a> {
                     )))
             );
         };
+
         let mut body = String::new();
-        let parser = utils::create_new_pull_down_parser(&ch.content);
+
+        if self.config.epub_version == Some(3) && self.config.footnote_backrefs {
+            body.push_str(&self.render_with_footnote_backrefs(chapter_dir, ch));
+        } else {
+            let parser = utils::create_new_pull_down_parser(&ch.content);
+            let mut quote_converter = EventQuoteConverter::new(self.config.curly_quotes);
+            let ch_depth = chapter_dir.components().count();
+
+            // create 'Remote Assets' copy to be processed by AssetLinkFilter
+            let mut remote_assets: HashMap<String, Asset> = HashMap::new();
+            for (key, value) in self.assets.clone().into_iter() {
+                trace!("{} / {:?}", key, &value);
+                if let AssetKind::Remote(ref remote_url) = value.source {
+                    trace!(
+                        "Adding remote_assets = '{}' / {:?}",
+                        remote_url.to_string(),
+                        &value
+                    );
+                    remote_assets.insert(remote_url.to_string(), value);
+                }
+                /* else if let AssetKind::Local(ref _local_path) = value.source {
+                let relative_path = value.filename.to_str().unwrap();
+                remote_assets.insert(String::from(relative_path), value);
+                } */
+            }
+            let asset_link_filter = AssetLinkFilter::new(&remote_assets, ch_depth);
+            let events = parser
+                .map(|event| quote_converter.convert(event))
+                .map(|event| asset_link_filter.apply(event));
+            trace!("Found Rendering events map = [{:?}]", &events);
+
+            html::push_html(&mut body, events);
+        }
+
+        trace!("Chapter content after Events processing = [{:?}]", body);
+
+        let stylesheet_path = chapter_dir
+            .components()
+            .map(|_| "..")
+            .chain(iter::once("stylesheet.css"))
+            .collect::<Vec<_>>()
+            .join("/");
+
+        let epub_version_3 = self.config.epub_version == Some(3);
+
+        let ctx = json!({
+            "epub_version_3": epub_version_3,
+            "title": ch.name,
+            "body": body,
+            "stylesheet": stylesheet_path
+        });
+
+        self.hbs.render("index", &ctx)
+    }
+
+    fn render_with_footnote_backrefs(&self, chapter_dir: &Path, ch: &Chapter) -> String {
```

Comment:
I believe extracting closures at L304 & L385 to functions would improve readability. And extracting link filter logic to a function may reduce repetition.

Reply:
I managed to extract the one that was at L304 with the compiler messages helping me through it, but the other one gave me errors that I don't know how to deal with.